packages/web3-common/src/web3_base_provider.ts

```diff
@@ -0,0 +1,79 @@
+import {
+	JSONRPCPayload,
+	JSONRPCRequest,
+	JSONRPCResponseWithError,
+	JSONRPCResponseWithResult,
+	JSONRPCResult,
+} from './types';
+
+export interface ProviderMessage<T = JSONRPCResult> {
+	type: string;
+	data: T;
+}
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#connectivity
+export type Web3BaseProviderStatus = 'connecting' | 'connected' | 'disconnected';
+export type Web3BaseProviderCallback<T = JSONRPCResult> = (
+	error: Error | null,
+	result?: ProviderMessage<T>,
+) => void;
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#provider-errors
+export const JSONRPC_ERR_REJECTED_REQUEST = 4001;
+export const JSONRPC_ERR_UNAUTHORIZED = 4001;
+export const JSONRPC_ERR_UNSUPPORTED_METHOD = 4200;
+export const JSONRPC_ERR_DISCONNECTED = 4900;
+export const JSONRPC_ERR_CHAIN_DISCONNECTED = 4901;
+
+// Provider interface compatible with EIP-1193
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md
+export abstract class Web3BaseProvider {
+	// TODO: For legacy support, should be deprecated and removed in favor of `request`
+	public send<T = JSONRPCResult, T2 = unknown[]>(
+		payload: JSONRPCPayload<T2>,
+		callback: (
+			error?: JSONRPCResponseWithError<T>,
+			result?: JSONRPCResponseWithResult<T>,
+		) => void,
+	): void {
+		this.request<T, T2>({ method: payload.method, params: payload.params })
+			.then(d =>
+				callback(undefined, { result: d, id: payload.id ?? 0, jsonrpc: payload.jsonrpc }),
+			)
+			.catch(e => callback(e, undefined));
+	}
+
+	abstract getStatus(): Web3BaseProviderStatus;
+	abstract supportsSubscriptions(): boolean;
+
+	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#request
+	abstract request<T = JSONRPCResult, T2 = unknown[]>(request: JSONRPCRequest<T2>): Promise<T>;
+
+	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#events
+	abstract on<T = JSONRPCResult>(
+		type: 'message' | 'disconnect' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void;
+	abstract on(
+		type: 'connect' | 'chainChanged',
+		callback: Web3BaseProviderCallback<{
+			readonly [key: string]: unknown;
+			readonly chainId: string;
+		}>,
+	): void;
+	abstract on(
+		type: 'accountsChanged',
+		callback: Web3BaseProviderCallback<{
+			readonly [key: string]: unknown;
+			readonly accountsChanged: string[];
+		}>,
+	): void;
```

Comment:
`.on` should only exist for web sockets and EIP1193 providers, no?

Reply:
No on should exists on abstract class as its very basic behaviour of any provider. If some provider implementing this abstract class does not provide this behaviour it should still implement and throw `NotSupportedError`.